core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Map;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+class UpdateSnapshotReferencesOperation implements PendingUpdate<Map<String, SnapshotRef>> {
```

Comment:
Can you add a `TODO` comment to move the operations in `SetSnapshotOperation` to this class? I don't think that needs to be done as part of this commit, but we should do it eventually.

That operation handles `setCurrentSnapshot`, `rollbackToTime`, and `rollbackTo`, which are all operations that this class could handle by updating the set of refs. Also, `rollbackTo` could have a branch version, `rollback(String name, long snapshotId)` that would validate the ref is a branch and validate that the snapshot is an ancestor before updating the ref.

Reply:
Yup! I'm in agreement there, all the snapshot operations can be applied within the context of refs as well. I'll add a class level comment